extensions/librdkafka/rdkafka_utils.cpp

```diff
@@ -0,0 +1,115 @@
+  }
+}
+
+void print_topics_list(std::shared_ptr<logging::Logger> logger, rd_kafka_topic_partition_list_t* kf_topic_partition_list) {
+  for (std::size_t i = 0; i < kf_topic_partition_list->cnt; ++i) {
+    logger->log_debug("kf_topic_partition_list: topic: %s, partition: %d, offset:%lld]",
+    kf_topic_partition_list->elems[i].topic, kf_topic_partition_list->elems[i].partition, kf_topic_partition_list->elems[i].offset);
+  }
+}
+
+void print_kafka_message(const rd_kafka_message_t* rkmessage, const std::shared_ptr<logging::Logger>& logger) {
+  if (RD_KAFKA_RESP_ERR_NO_ERROR != rkmessage->err) {
+    const std::string error_msg = "ConsumeKafka: received error message from broker. Librdkafka error msg: " + std::string(rd_kafka_err2str(rkmessage->err));
+    throw minifi::Exception(ExceptionType::PROCESSOR_EXCEPTION, error_msg);
+  }
+  std::string topicName = rd_kafka_topic_name(rkmessage->rkt);
+  std::string message(reinterpret_cast<char*>(rkmessage->payload), rkmessage->len);
+  const char* key = reinterpret_cast<const char*>(rkmessage->key);
+  const std::size_t key_len = rkmessage->key_len;
+  rd_kafka_timestamp_type_t tstype;
+  int64_t timestamp;
+  timestamp = rd_kafka_message_timestamp(rkmessage, &tstype);
+  const char *tsname = "?";
+  if (tstype != RD_KAFKA_TIMESTAMP_NOT_AVAILABLE) {
+    if (tstype == RD_KAFKA_TIMESTAMP_CREATE_TIME) {
+      tsname = "create time";
+    } else if (tstype == RD_KAFKA_TIMESTAMP_LOG_APPEND_TIME) {
+      tsname = "log append time";
+    }
+  }
+  const int64_t seconds_since_timestamp = timestamp ? static_cast<int64_t>(time(NULL)) - static_cast<int64_t>(timestamp / 1000) : 0;
+
+  std::string headers_as_string;
+  rd_kafka_headers_t* hdrs;
+  const rd_kafka_resp_err_t get_header_response = rd_kafka_message_headers(rkmessage, &hdrs);
+  if (RD_KAFKA_RESP_ERR_NO_ERROR == get_header_response) {
+    std::vector<std::string> header_list;
+    kafka_headers_for_each(hdrs, [&] (const std::string& key, const std::string& val) { header_list.emplace_back(key + ": " + val); });
+    headers_as_string = StringUtils::join(", ", header_list);
+  } else if (RD_KAFKA_RESP_ERR__NOENT != get_header_response) {
+    logger->log_error("Failed to fetch message headers: %d: %s", rd_kafka_last_error(), rd_kafka_err2str(rd_kafka_last_error()));
+  }
+
+  std::string message_as_string;
+  message_as_string += "[Topic](" + topicName + "), ";
+  message_as_string += "[Key](" + (key != nullptr ? std::string(key, key_len) : std::string("[None]")) + "), ";
+  message_as_string += "[Offset](" +  std::to_string(rkmessage->offset) + "), ";
+  message_as_string += "[Message Length](" + std::to_string(rkmessage->len) + "), ";
+  message_as_string += "[Timestamp](" + std::string(tsname) + " " + std::to_string(timestamp) + " (" + std::to_string(seconds_since_timestamp) + " s ago)), ";
+  message_as_string += "[Headers](";
+  message_as_string += headers_as_string + "\n";
```

Comment:
Should this `"\n"` be `"), "`?  https://xkcd.com/859/

Reply:
Oh no. What have I done.